src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -168,6 +167,22 @@ public UnfilteredRowIterator queryStorage(PrimaryKey key, ReadExecutionControlle
         return partition.queryMemtableAndDisk(cfs, executionController);
     }
 
+    private static Runnable getIndexReleaser(Set<SSTableIndex> referencedIndexes)
+    {
+        return new Runnable()
+        {
+            boolean closed;
+            @Override
+            public void run()
+            {
+                if (closed)
+                    return;
+                closed = true;
+                referencedIndexes.forEach(SSTableIndex::releaseQuietly);
+            }
+        };
+    }
```

Comment:
I seem to remember us trying to explicitly make sure we only released the SSTable indexes once in CASSANDRA-19428. Did we miss a case, or is there something specific to this patch that made this check necessary?

Reply:
Close is on iterators is supposed to be idempotent. It's not in this case and you can end up double freeing the sstable index breaking reads. The code in `TxnNamedRead` was closing the inner and wrapping iterator so close ended up being called twice revealing the problem.

Reply:
The iterator `executeLocally()` produces is closed twice? i.e. The SAI internal code still closes things at the proper level, but `TxnNamedRead` is closing both? If so, making the actual site of the closure logic in `QueryController` enforce idempotence makes sense.

Reply:
I wasn't able to reproduce the double close on the `KeyRangeIterator` w/ `AccordSingleNodeSAITest`. Is there an easier way? `BaseIterator` (which `FilteredPartitions` extends) does seem to check for double closure already. Sorry for the stream of consciousness...

Reply:
This commit also updates `TxnNamedRead` to not do the double close so you have to revert that as well in order to get the test to hang.

Reply:
I see. So we're just trying to make sure that can't happen in the future. In that case, we might want to consider just enforcing single closure in `KeyRangeIterator#close()`. Might be slightly less invasive...up to you.